### PR TITLE
Submittal service: the first view of controllers

### PR DESCRIPTION
**Pull request by abelan**
there is no swagger/tests/docs > I would like to be sure I'm on the right way

### CODE REVIEW

File: submittal/build.gradle [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+version '0.0.1-SNAPSHOT'
+sourceCompatibility = 1.8
+
+repositories {
+    mavenCentral()
+}
+
+dependencies {
+    implementation 'org.springframework.boot:spring-boot-starter-web',
+            'org.springframework.boot:spring-boot-starter-data-jpa'
```
Comment (xpinjection) [inline submittal/build.gradle:31]: We don't need JPA at that stage.

File: submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+/**
+ * @author M-ABL
+ */
+@RestController
+@RequestMapping(value = "/api/v1/confirm",
+    consumes = APPLICATION_JSON_UTF8_VALUE,
+    produces = APPLICATION_JSON_UTF8_VALUE)
+public class ConfirmationController
+{
+  private final SpeakerService speakerService;
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java:31]: We don't need service at that stage.

File: submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+
+  @Autowired
+  ConfirmationController(final SpeakerService speakerService)
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java]: AllArgsConstructor could replace this

File: submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  {
+    this.speakerService = speakerService;
+  }
+
+
+  @GetMapping(path = "/registration")
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java]: On GET request there must be no active business logic, just reading. And resource URL must be identifiable.

File: submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  public RegistrationConfirmedResponse confirmRegistration(@RequestParam final String key)
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java]: public is not needed

File: submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+
+/**
+ * @author M-ABL
+ */
+@RestController
+@RequestMapping(value = "/api/v1/confirm",
+    consumes = APPLICATION_JSON_UTF8_VALUE,
+    produces = APPLICATION_JSON_UTF8_VALUE)
+public class ConfirmationController
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java:29]: There is no API documentation

Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java:29]: There are no validation rules

File: submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  {
+    return RegistrationConfirmedResponse.builder()
+        .speakerId(speakerService.finishRegistration(key).getSpeakerId())
+        .build();
+    // TODO: what's next? Invoke Conference MS directly in order to submit topics?
+  }
+
+
+  @Builder
+  private static class RegistrationConfirmedResponse
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java]: Builder is useless for such DTO, this is Value Object

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+import static org.springframework.http.MediaType.APPLICATION_JSON_UTF8_VALUE;
+
+
+/**
+ * @author M-ABL
+ */
+// TODO: setup security. Let's assume a filter retrieves a speaker's id by email/password and puts the id to the session.
+@RestController
+@RequestMapping(value = "/api/v1/speaker",
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: Common prefix could be moved to common parent class.

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    consumes = APPLICATION_JSON_UTF8_VALUE,
+    produces = APPLICATION_JSON_UTF8_VALUE)
+public class SpeakerController
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java:45]: The same set of comments as for another controller.

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+
+  @Autowired
+  SpeakerController(final SpeakerService speakerService)
+  {
+    this.speakerService = speakerService;
+  }
+
+
+  @PostMapping(path = "/register")
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java:66]: POST requests must be on URL .../speakers

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+  public SpeakerResponse register(@RequestBody final SpeakerRequest speaker)
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: public and final are useless here

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+        speaker.email, speaker.password, speaker.personalData, speaker.topics, speaker.conferenceId);
+    return SpeakerResponse.builder()
+        .email(speaker.email)
+        .confirmationKey(key)
+        .build();
+  }
+
+
+  @PostMapping(path = "/topics")
+  public TopicResponse proposeTopics(@SessionAttribute("speakerId") long speakerId,
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: Session is forbidden for usage in REST API, move to URL path

Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: public and final are useless here

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+        .proposeTopics(speakerId, topicRequest.topics);
+    return TopicResponse.builder()
+        .speakerId(result.getSpeakerId())
+        .topics(result.getData())
+        .build();
+  }
+
+
+  @PutMapping(path = "/personal")
+  public void modifyPersonalData(@SessionAttribute("speakerId") long speakerId,
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: Again URL must be like .../speakers/{speakerId}/personal

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+                                 @RequestBody final PersonalInfoDto personalData)
+  {
+    speakerService.modifyPersonalData(speakerId, personalData);
+  }
+
+
+  @Builder
+  private static class TopicRequest
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java:113]: Builder is useless for all those DTOs

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java (hunk 12 above)
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: On POST some kind of resource ID is returned

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+    private PersonalInfoDto personalData;
+    private List<TopicDto> topics;
+    private long conferenceId;
+  }
+
+  @Builder
+  private static class SpeakerResponse
+  {
+    private String email;
+    private String confirmationKey;
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: What is this?

File: submittal/src/main/java/com/kyriba/submittal/dao/SpeakerDao.java [hunk 16]
```diff
@@ -0,0 +1,12 @@
+package com.kyriba.submittal.dao;
+
+import org.springframework.stereotype.Repository;
+
+
+/**
+ * @author M-ABL
+ */
+@Repository
+public class SpeakerDao
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/dao/SpeakerDao.java]: It is not needed at that stage

File: submittal/src/main/java/com/kyriba/submittal/dao/TopicDao.java [hunk 17]
```diff
@@ -0,0 +1,12 @@
+package com.kyriba.submittal.dao;
+
+import org.springframework.stereotype.Repository;
+
+
+/**
+ * @author M-ABL
+ */
+@Repository
+public class TopicDao
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/dao/TopicDao.java]: It is not needed at that stage

File: submittal/src/main/java/com/kyriba/submittal/domain/Speaker.java [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+import java.util.List;
+
+
+/**
+ * @author M-ABL
+ */
+public class Speaker
+{
+  private long id;
+  private String email;
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/domain/Speaker.java:28]: Probably properties could be grouped in logical classes like Address, ContactData, etc.

File: submittal/src/main/java/com/kyriba/submittal/domain/dto/JobInfoDto.java [hunk 19]
```diff
@@ -0,0 +1,60 @@
+package com.kyriba.submittal.domain.dto;
+
+/**
+ * @author M-ABL
+ */
+public class JobInfoDto
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/domain/dto/JobInfoDto.java]: Getters and setters could be replaced with Lombok Data annotation

File: submittal/src/main/java/com/kyriba/submittal/domain/dto/PersonalInfoDto.java [hunk 20]
```diff
@@ -0,0 +1,13 @@
+package com.kyriba.submittal.domain.dto;
+
+/**
+ * @author M-ABL
+ */
+public class PersonalInfoDto
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/domain/dto/PersonalInfoDto.java]: Common parts could be reused after extraction from Speaker.

File: submittal/src/main/java/com/kyriba/submittal/domain/dto/SpeakerDataDto.java [hunk 21]
```diff
@@ -0,0 +1,31 @@
+package com.kyriba.submittal.domain.dto;
+
+/**
+ * @author M-ABL
+ */
+public class SpeakerDataDto<T>
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/domain/dto/SpeakerDataDto.java]: Very strange class

File: submittal/src/main/java/com/kyriba/submittal/service/KeyGenerator.java [hunk 22]
```diff
@@ -0,0 +1,20 @@
+package com.kyriba.submittal.service;
+
+import org.springframework.stereotype.Component;
+
+
+/**
+ * @author M-ABL
+ */
+@Component
+public class KeyGenerator
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/service/KeyGenerator.java:10]: It is not needed at that stage

File: submittal/src/main/java/com/kyriba/submittal/service/SpeakerService.java [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+import com.kyriba.submittal.domain.dto.SpeakerDataDto;
+import com.kyriba.submittal.domain.dto.TopicDto;
+
+import java.util.List;
+
+
+/**
+ * @author M-ABL
+ */
+public interface SpeakerService
```
Comment (xpinjection) [inline submittal/src/main/java/com/kyriba/submittal/service/SpeakerService.java:16]: It is not needed at that stage

File: submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java (hunk 3 above)
Comment (abelan) [reply submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java]: I know but is it mandatory to use lombok?

File: submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java (hunk 4 above)
Comment (abelan) [reply submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java]: and what's the best option for the confirmation request? POST?

File: submittal/src/main/java/com/kyriba/submittal/domain/dto/SpeakerDataDto.java (hunk 21 above)
Comment (abelan) [reply submittal/src/main/java/com/kyriba/submittal/domain/dto/SpeakerDataDto.java]: the idea is to return SpeakerId and some extra data

File: submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java (hunk 2 above)
Comment (abelan) [reply submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java:31]: this is just an interface

File: submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java (hunk 7 above)
Comment (abelan) [reply submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java]: please explain. I use the builder in order to set values and return as response

File: submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java (hunk 3 above)
Comment (xpinjection) [reply submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java]: For dependency injection it is common pattern if Lombok is used at all.

File: submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java (hunk 4 above)
Comment (xpinjection) [reply submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java]: POST and PUT are modification commands, GET is not and idempotent.

File: submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java (hunk 2 above)
Comment (xpinjection) [reply submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java:31]: We will review on the second module how to create it, where to store and how to manage.

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java (hunk 11 above)
Comment (abelan) [reply submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: I always use 'final' for arguments

File: submittal/src/main/java/com/kyriba/submittal/domain/dto/SpeakerDataDto.java (hunk 21 above)
Comment (xpinjection) [reply submittal/src/main/java/com/kyriba/submittal/domain/dto/SpeakerDataDto.java]: Usually DTOs are created as simple as possible and composition with generics is not used (tools work bad, not clear code, etc.).

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java (hunk 12 above)
Comment (abelan) [reply submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: Will we apply the security later?

File: submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java (hunk 7 above)
Comment (xpinjection) [reply submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java]: Builder pattern is used when class has many fields and dynamic creation options are needed. With single param it is easier to use Value annotation and constructor for creation.

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java (hunk 12 above)
Comment (abelan) [reply submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: TopicResponse contains id of a speaker and id of each created topic. What else I missed?

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java (hunk 11 above)
Comment (xpinjection) [reply submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: Controllers are storage for meta-information, so all unnecessary keywords may be removed for simplification

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java (hunk 12 above)
Comment (xpinjection) [reply submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: Yes, and definitely not with Session. :)

Comment (xpinjection) [reply submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: It is expected to return only ID of just created Topic.

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java (hunk 15 above)
Comment (abelan) [reply submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: the user has to confirm the registration. He'll get an email with a key. With a specific link a new request to ConfirmationController is performed

File: submittal/src/main/java/com/kyriba/submittal/domain/Speaker.java (hunk 18 above)
Comment (abelan) [reply submittal/src/main/java/com/kyriba/submittal/domain/Speaker.java:28]: email here is more like a login

File: submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java (hunk 2 above)
Comment (abelan) [reply submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java:31]: should I remove such classes from this PR?

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java (hunk 8 above)
Comment (abelan) [reply submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: Should we specify a name of a service in the url? e.g. /api/v1/submittal and then specific path? Or submittal is redundant in this case?

File: submittal/src/main/java/com/kyriba/submittal/domain/Speaker.java (hunk 18 above)
Comment (xpinjection) [reply submittal/src/main/java/com/kyriba/submittal/domain/Speaker.java:28]: ok, but comment still actual for other fields

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java (hunk 8 above)
Comment (xpinjection) [reply submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: In REST style API we don't specify service name in URL. Instead we operate only with resource names.

File: submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java (hunk 2 above)
Comment (xpinjection) [reply submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java:31]: This is not critical.

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java (hunk 15 above)
Comment (xpinjection) [reply submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: OK, I got the idea.

File: submittal/src/main/java/com/kyriba/submittal/domain/Speaker.java (hunk 18 above)
Comment (abelan) [reply submittal/src/main/java/com/kyriba/submittal/domain/Speaker.java:28]: well, I have e.g. domain.dto.JobInfoDto. following your proposal I should also create domain.JobInfo domain object. What's the difference?

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java (hunk 8 above)
Comment (abelan) [reply submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: what's a common part in this case? '/api/v1' ? is it OK when the first controller has 'v1' version and the second 'v2'? If so, then a version is not a common part. Seems only consumes and produces can be moved. But a parent class looks empty then.

File: submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java (hunk 4 above)
Comment (abelan) [reply submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java]: can we pass a confirmation key as a RequestParam for POST method? or it's better to put it into the body?

File: submittal/src/main/java/com/kyriba/submittal/domain/dto/JobInfoDto.java (hunk 19 above)
Comment (abelan) [reply submittal/src/main/java/com/kyriba/submittal/domain/dto/JobInfoDto.java]: the same for 'domain' classes?

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java (hunk 8 above)
Comment (xpinjection) [reply submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: Usually whole microservice has the same version of API.

File: submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java (hunk 4 above)
Comment (xpinjection) [reply submittal/src/main/java/com/kyriba/submittal/api/ConfirmationController.java]: Body is the primary source of information for POST command. Sometimes headers are also used.

File: submittal/src/main/java/com/kyriba/submittal/domain/dto/JobInfoDto.java (hunk 19 above)
Comment (xpinjection) [reply submittal/src/main/java/com/kyriba/submittal/domain/dto/JobInfoDto.java]: This topic will be discussed tomorrow. :)

File: submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java (hunk 8 above)
Comment (abelan) [reply submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: please clarify one more time: the common part is '/api/v1' and for each method: '/speakers/...' ?

Comment (xpinjection) [reply submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: Common part for all controllers is /api/v1, common part for SpeakerController is /speakers.

Comment (abelan) [reply submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: noob question then: how to extend RequestMapping annotation in order to combine a path ('api/v1' + '/speakers')? As I know, the path is overridden but not merged

Comment (xpinjection) [reply submittal/src/main/java/com/kyriba/submittal/api/SpeakerController.java]: One common approach is to use config param for API prefix at class level like @RequestMapping(path = "${api.version.path}/users"). There are some other options: https://stackoverflow.com/questions/28006501/how-to-specify-prefix-for-all-controllers-in-spring-boot

**Pull request merged**